File: neural/python/losses.py
import numpy as np


def MSE(labels, outputs, derivative=False):
    # if type(labels) == np.ndarray or type(outputs) == np.ndarray:
    #     labels_shape, outputs_shape = np.product(labels.shape), np.product(outputs.shape)

    #     if labels_shape > outputs_shape:
    #         outputs = utils.pad_matrix(outputs, labels)
    #     else:
    #         labels = utils.pad_matrix(labels, outputs)

    if derivative:
        return np.mean(2 * (outputs - labels))
    
    return np.mean((outputs - labels) ** 2)

def categorical_crossentropy(labels, logits, derivative=False):
    if derivative:
        return np.mean(-(1 / logits).dot(labels))

    return np.mean(-np.log(logits).dot(labels))

File: neural/cpp/Neural/Neural/src/losses.cc
#include "losses.h"

_differentiable_scalar Neural::Losses::MSE(Eigen::VectorXd labels, Eigen::MatrixXd outputs)
{
	_differentiable_scalar loss;

	loss.function = (outputs.array().colwise() - labels.array()).square().mean();
	loss.derivative = (2 * (outputs.array().colwise() - labels.array())).mean();

	return loss;
}

_differentiable_scalar Neural::Losses::CategoricalCrossentropy(Eigen::MatrixXd labels, Eigen::MatrixXd logits)
{
	_differentiable_scalar loss;

	loss.function = ((-logits).array().log()/* * labels.array()*/).mean();
	loss.derivative = ((-labels).array() / logits.array()).mean();

	return loss;
}